### ISSUE TITLE
How can I add an expiration time with access token  ?

### ISSUE THREAD

**Issue by rahulranjan07** [ahopkins/sanic-jwt#50]
I'm trying to implement a logic on client side to ask for a new token whenever the previous one expires by storing the expiry time on client side. Is it possible to add an expiration time with the `/auth` end point ?

**Comment by ahopkins** [comment]
So I understand the question, you want to hit the endpoint and receive back a response that looks something like this:

    {
        “access_token”: “<TOKEN>”,
        “expiration”: “<TIMESTAMP”,
    }

This is a requested feature. See #13 and #14. Unfortunately, there is not yet an implementation for this. Although, I have been working on a solution on a separate project. I will probably add this ability in the near future.

Until that point, one option would be to extract the expiration time of the access token by the client from the payload. 

Here is an example of how I did something similar in my JavaScript front end client using [momentjs](http://momentjs.com/) and [jwt-decode](http://momentjs.com/).

    import jwt_decode from ‘jwt-decode’
    import moment from ‘moment’

    const token = response.data.access_token
    const payload = jwt_decode(token)
    const expires = moment.unix(payload.exp)
    console.log('expires', expires)

**Comment by vltr** [comment]
Isn't the expiration time (`exp`) already included into jwt? The main problem here would be the client to "presume" the state of something that's only genuine to the server (in this case, the validity of the token).

But why "presume"? Trying to "guess" if the token is still valid can lead you to lots of problems (almost) **unrelated** to jwt:
- The user internet connection (speed, latency);
- The (probable) difference between server and client timezones (and / or the clock itself);
- Security concerns (like if you're using a secret, either you won't be able to decode in your client or it'll become useless if you expose it).

IMO, one "not presumable" way for doing this is a simple heartbeat to `/auth/me` (or any other endpoint) and, if the token become indeed expired (issued by the server), then all you have to do is create a new one (using the `refresh_token`) before creating any other request.

You can also wrap your calls with this same implementation prior to any request.

**Comment by ahopkins** [comment]
@vltr all good points.

Yes, the `exp` will automatically be added to every payload. The one exception is if you make your access tokens non-expirable. **HIGHLY** discouraged since you would effectively have no way to invalidate the token (except maybe changing your secret).

Also, that is certainly one point of the `/auth/verify` endpoint. In a Javascript frontend, I capture the `exp` and then subtract some amount of time to prompt the user to see if they want to keep the session alive. If yes, then go get a refresh token.

```javascript
    checkForTimeout () {
        const {expires} = this.props
        if (expires) {
            const warningNotice = 5
            const warningTime = expires.subtract(warningNotice, 'minutes')
            const secondsToAlert = warningTime.diff(moment())
            setTimeout(() => {
                this.props.notify('warning', {
                    title: "Your session is about to expire",
                    message: `In ${warningNotice} minutes, your session will expire. Do you wish to continue?`,
                    autoDismiss: 0,
                    action: {
                        label: 'Continue',
                        callback: function () {
                            this.props.refreshAccessToken()
                        }.bind(this)
                    }
                })
            }, secondsToAlert)
        }
    }
```

Wrapping all of your calls with `/auth/verify` seems overkill for most web applications. But, then again this is the business logic where EVERY application, whether web based or not, will have different requirements, and therefore need to do something different. Which is why to the extent possible flexibility is the key.

**Comment by vltr** [comment]
> The one exception is if you make your access tokens non-expirable. **HIGHLY** discouraged since [...]

Indeed, non-expirable tokens shouldn't even exist! :sweat_smile:

> Also, that is certainly one point of the `/auth/verify` endpoint. In a Javascript frontend, I capture the `exp` and then subtract some amount of time to prompt the user to see if they want to keep the session alive. If yes, then go get a refresh token.

Yes, having a "decreased" expiration time would help mitigate the possible problems I mentioned, but that's just for web applications (that doesn't have a "sleeping mode"). If we talk about other potential consumers, specially mobile apps connected using 3G, 4G ... Then things starts to get complicated (which is one of my cases).

> Wrapping all of your calls with `/auth/verify` seems overkill for most web applications.

Yeap, I agree, specially if the backend and the frontend is (technically) developed by the same team, so you end up knowing how much time the token takes to expire. The problem is when you don't, as you kind of mentioned.

> Which is why to the extent possible flexibility is the key.

Perhaps add the possibility to send the expiration timestamp with the `/auth` response, either using the Unix (converted to UTC) or [ISO 8601](https://xkcd.com/1179/) (extended) formats, being the later preferable. But ... Isn't that something that could be addressed using an implementation for the response body similar to `SANIC_JWT_HANDLER_PAYLOAD_EXTEND`?

**Comment by ahopkins** [comment]
Good point. But, on the expiration issue, I thought the same. And then I realized that I had a use case where I needed an API that would eventually be consumable from a frontend client, but at the time was only being used between two machines. The non-expirable key actually did make sense as a potential solution. In the end I decided against it, but I could at least see the existence of a use case for it.
